### PR TITLE
Remove mem::transmute used in Box<str> conversions

### PR DESCRIPTION
**Pull request by nvzqz**
Given that https://github.com/rust-lang/rust/pull/44877 is failing, I decided to make a separate PR. This is done with the same motivation: to avoid `mem::transmute`-ing non `#[repr(C)]` types.

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by nvzqz** [comment]
I may add some more similar changes for `OsStr` and `CStr` APIs before this gets merged, if that's ok.

**Comment by alexcrichton** [comment]
Sure! Feel free to just ping me when that happens

File: src/libstd/ffi/c_str.rs [hunk 1]
```diff
@@ -312,7 +312,7 @@ impl CString {
     pub unsafe fn from_raw(ptr: *mut c_char) -> CString {
         let len = libc::strlen(ptr) + 1; // Including the NUL byte
         let slice = slice::from_raw_parts(ptr, len as usize);
-        CString { inner: mem::transmute(slice) }
+        CString { inner: Box::from_raw(slice as *mut [c_char] as *mut [u8]) }
```
Comment (kennytm) [inline src/libstd/ffi/c_str.rs]: Use `slice::from_raw_parts_mut` to get a `&mut [i8]`.

```
[00:02:36] error[E0606]: casting `&[i8]` as `*mut [i8]` is invalid
[00:02:36]    --> /checkout/src/libstd/ffi/c_str.rs:315:40
[00:02:36]     |
[00:02:36] 315 |         CString { inner: Box::from_raw(slice as *mut [c_char] as *mut [u8]) }
[00:02:36]     |                                        ^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by nvzqz** [comment]
@alexcrichton this should be good to go once cb2a656 passes.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by kennytm** [comment]
@nvzqz The bors error is legitimate.

**Comment by nvzqz** [comment]
@kennytm I committed another change that should fix the error on appveyor, so I was trying to get bors to try again with the new changes in this PR

**Comment by kennytm** [comment]
@nvzqz retry only works on approved commits, if you create a new commit, someone will need to r+ it again 🙂

**Comment by nvzqz** [comment]
I didn't know that. Thanks!

**Pull request merged**